controllers/kserve_inferenceservice_controller.go

```diff
@@ -468,6 +470,25 @@ func (r *OpenshiftInferenceServiceReconciler) createOrUpdateMetricsServiceMonito
 	return nil
 }
 
+func (r *OpenshiftInferenceServiceReconciler) DeleteKserveInferenceServiceRoute(ctx context.Context, inferenceService *kservev1beta1.InferenceService) error {
+	if r.isDeploymentModeForIsvcModelMesh(inferenceService) {
+		return nil
+	}
```

Comment:
Perhaps, this check shouldn't be done. It may be that the ISVC was switched from KServe mode to ModelMesh mode. 

In general, if there is an owned route which should not exist, the controller should delete it regardless of the deployment mode.

Reply:
This route doesn't have owner reference because its created in `istio-system` namespace. Owner references could only be assigned within same namespaces thats why I have to explicitly write its deletion on deletion of isvc.

Reply:
Think of "owned", generally speaking, as a route that was created by the controller (not about the annotations).

I agree that explicit deletion is needed.

The question is around the conditional: `if r.isDeploymentModeForIsvcModelMesh(inferenceService)`.

I haven't tried the changes. But my perception is that given an `InferenceService` **without** the `serving.kserve.io/deploymentMode` annotation, the controller would correctly create a route. If later, the `serving.kserve.io/deploymentMode: ModelMesh` is added to this same InferenceService, this conditional will bypass deletion and the OpenShift Route will still exist as a remainder (which may be unwanted).

Reply:
@israel-hdez Do we support such type of transition from Kserve to ModelMesh or vice versa?

Reply:
> Do we support such type of transition from Kserve to ModelMesh or vice versa?
It depends on if KServe or ModelMesh block the edit via some Validating WebHook.

I can say that ModelMesh does not block the edit. But I'm unsure about KServe.

If the edit is allowed, it becomes _desired state_ and `odh-model-controller` should support it.

> I believe there would be multiple place that need this type of handling.
Probably yes.

> Could I create seperate ticket for it?
You can, although I'd think it should take priority. It is easy to forget the annotation for ModelMesh. If `odh-model-controller` is released as is, it would be seen as a bug.

Reply:
I had created below ticket for it with Critical Priority.
https://github.com/opendatahub-io/odh-model-controller/issues/88